src/main/webapp/signup.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * @typedef SignUpData
+ * @property {string} gitHubAuthToken
+ * @property {string[]} interestTags
+ */
+import {gitHubAuthorizer} from './authorization.js';
+import {
+  standardizeFetchErrors,
+  makeRelativeUrlAbsolute,
+} from './js/fetch_handler.js';
+
+const AFTER_SIGNUP_REDIRECT = '/dashboard.html';
+
+/**
+ * Initializes the sign up form.
+ */
+function initSignupForm() {
+  const signupForm = document.getElementById('signup-form');
+  signupForm.addEventListener('submit', submitSignup);
+
+  signupForm.elements['github-link-button'].addEventListener(
+      'click', handleGitHubLink);
+  gitHubAuthorizer
+      .getFirebase().auth().onAuthStateChanged(handleAuthStateChanged);
+  // Handle auth state changed on page load in case of existing authentication.
+  handleAuthStateChanged(gitHubAuthorizer.getUser());
+}
+
+/**
+ * Handles the click event for the GitHub linking button.
+ * @param {Event} e The button click event.
+ */
+function handleGitHubLink(e) {
+  e.preventDefault();
+
+  gitHubAuthorizer.toggleSignIn().catch((error) => {
+    console.error(error);
+    alert('Encountered an error with GitHub authentication.'
+        + ' Please try again later.');
+  });
+}
+
+/**
+ * Updates the signup UI when the state of user GitHub authentication changes.
+ * @param {Firebase.User} user
+ */
+function handleAuthStateChanged(user) {
+  const signupForm = document.getElementById('signup-form');
+  const submitButton = signupForm.elements['submit-button'];
+  const gitHubLinkButton = signupForm.elements['github-link-button'];
+  if (user) {
+    submitButton.disabled = false;
+    gitHubLinkButton.innerText = 'Unlink Your GitHub Account';
+    gitHubLinkButton.classList.add('btn-success');
+    gitHubLinkButton.classList.remove('btn-purple');
+  } else {
+    submitButton.disabled = true;
+    gitHubLinkButton.innerText = 'Link Your GitHub Account';
+    gitHubLinkButton.classList.add('btn-purple');
+    gitHubLinkButton.classList.remove('btn-success');
+  }
+}
+
+/**
+ * Validates and submits a user signup request.
+ * @param {Event} e The signup form submit event.
+ */
+function submitSignup(e) {
+  e.preventDefault();
+
+  const signupForm = document.getElementById('signup-form');
+  const gitHubLinkButton = signupForm.elements['github-link-button'];
+  const submitButton = signupForm.elements['submit-button'];
+  gitHubLinkButton.disabled = true;
+  submitButton.disabled = true;
+
+  const signupBody = createSignupBody();
+  console.log('Body of signup request:');
+  console.log(signupBody);
+  if (signupBody) {
+    const signupRequest = createSignupRequest(signupBody);
```

Comment:
I think you could probably move `createSignupBody` within `createSignupRequest`. That way you wouldn't have to have an extra check to see if signupBody is not null. You could just check whether createSignupRequest succeeded or raised an error.